controller/src/main/java/io/pravega/controller/store/client/StoreClientFactory.java

```diff
@@ -121,38 +121,16 @@ static class ZKClientFactory implements ZookeeperFactory {
         @Override
         @Synchronized
         public ZooKeeper newZooKeeper(String connectString, int sessionTimeout, Watcher watcher, boolean canBeReadOnly) throws Exception {
-            // prevent creating a new client, stick to the same client created earlier
-            // this trick prevents curator from re-creating ZK client on session expiry
             if (client == null) {
                 Exceptions.checkNotNullOrEmpty(connectString, "connectString");
                 Preconditions.checkArgument(sessionTimeout > 0, "sessionTimeout should be a positive integer");
                 this.connectString = connectString;
                 this.sessionTimeout = sessionTimeout;
                 this.canBeReadOnly = canBeReadOnly;
-                this.client = new ZooKeeper(connectString, sessionTimeout, watcher, canBeReadOnly);
-            } else {
-                try {
-                    Preconditions.checkArgument(this.connectString.equals(connectString), "connectString differs");
-                    Preconditions.checkArgument(this.sessionTimeout == sessionTimeout, "sessionTimeout differs");
-                    Preconditions.checkArgument(this.canBeReadOnly == canBeReadOnly, "canBeReadOnly differs");
-                    this.client.register(watcher);
-                } catch (IllegalArgumentException e) {
-                    log.warn("Input argument for new ZooKeeper client ({}, {}, {}) changed with respect to existing client ({}, {}, {}).",
-                        connectString, sessionTimeout, canBeReadOnly, this.connectString, this.sessionTimeout, this.canBeReadOnly);
-                    closeClient(client);
-                }
             }
+            this.client = new ZooKeeper(this.connectString, this.sessionTimeout, watcher, this.canBeReadOnly);
```

Comment:
It looks like the Controller was keeping the same Curator client instance and probably this was for a reason. Do you know why? Do you know what are the potential implications of just re-creating the Curator client? (e.g., consistency, problems during Controller failovers, etc.)

Reply:
Checking the history says that this was introduced with the first commit itself. Obviously this is a good idea to re-use instance if available but we observed few instances where it was causing issues as explained as PR description. Moreover this is the same way we are using in Segment Stores as well and haven't found any issue with this approach. Ref:- https://github.com/pravega/pravega/blob/c9ffea2e6d1243e6a02fbc785f50130ae417150f/segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/ServiceStarter.java#L308 introduced with PR:- https://github.com/pravega/pravega/pull/3842 to handle similar issue on SSS.
I also do not see any potential implication of re-creating the Curator Client

Reply:
Question: aren't we closing this Zookeeper client? If there is a disconnection (i.e., connection lost exception in Curator) from Zookeeper, we just re-create the Zookeeper client without closing the previous one?

Reply:
right now we are overriding the previous one with new one. In fact we can actually close the previous one and then create a new one.